### PR TITLE
feat(indexer): add ENVIO_RPC_FALLBACK_URL_<chainId> for explicit fallback override

### PR DESCRIPTION
**Pull request by chapati23**
## Summary

- Adds `ENVIO_RPC_FALLBACK_URL_<chainId>` env var so the fallback RPC for `readContractWithBlockFallback` can be configured independently of the primary. The previous "fallback IS the hardcoded default, only used when primary is overridden" rule is preserved when the new env var is unset.
- Why: catch-up logs from today's deploy show **2,466 `[RPC_ARCHIVE_FALLBACK]` warnings in 30 minutes** — every single block-scoped Monad read fails on the QuickNode primary (40k-block pruning) and routes through `rpc2.monad.xyz`. Swapping their roles (primary = rpc2, fallback = QuickNode) eliminates the failed-primary round-trip during backfill while keeping QuickNode's higher rate limit available at head. Requires the new env var because today both primary and "default" can't be set independently.
- Compares fully-resolved URLs (with HyperRPC token applied) when checking `primary == fallback` to avoid creating a useless self-referencing fallback client.

## Test plan

- [x] 551 mocha tests pass (was 544 + 7 new fallback-resolution tests covering: no-override default, primary-overridden default, fallback-env-var override, primary == fallback no-op, unknown chain null, caching, bare-HyperRPC fallback null)
- [x] `pnpm typecheck` clean
- [x] `trunk fmt` + `trunk check` clean on changed files
- [ ] After merge: in Envio dashboard, **unset** `ENVIO_RPC_URL_143` (primary becomes default `rpc2.monad.xyz`) and **set** `ENVIO_RPC_FALLBACK_URL_143` to the existing QuickNode URL
- [ ] Redeploy and confirm: catch-up logs no longer show `[RPC_ARCHIVE_FALLBACK]` for Monad block-scoped reads (since rpc2 is now primary and serves directly without falling back)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Adjusts RPC fallback resolution used by `readContractWithBlockFallback`, which can change how the indexer behaves under rate limits and archive-depth misses across chains; mitigated by added edge-case handling and dedicated tests.
> 
> **Overview**
> Adds an explicit per-chain fallback RPC env var (`ENVIO_RPC_FALLBACK_URL_<chainId>`) so the indexer can configure the secondary `eth_call` endpoint independently from the primary RPC.
> 
> Updates `getFallbackRpcClient` to resolve fallback URLs from the new override (or from the hardcoded default only when the primary is overridden), cache `null` results, treat empty-string env vars as unset, and avoid self-referencing fallbacks by normalizing/compare resolved URLs (including trailing-slash differences) and rejecting bare HyperRPC endpoints.
> 
> Expands mocha coverage for fallback client resolution and updates `.env.example`, `README.md`, and `AGENTS.md` to document the new configuration and its caveats.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f0da341eca79ec9db9debf9a3d15d9c3c67db817. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chapati23** [comment]
@claude review

**Comment by chapati23** [comment]
cursor review

**Comment by chapati23** [comment]
@claude — addressed all 3 nits in c74dcf0. Summary:

- **`null` result is never cached** → Fixed in c74dcf0: `fallbackRpcClients` Map's value type is now `ReturnType<typeof createPublicClient> | null`, and every early-return path (unknown chain, bare HyperRPC, primary == fallback) writes `null` to the cache before returning. Subsequent calls now short-circuit on `has()` instead of recomputing. Also dropped the dead `?? null` per your suggestion.
- **Trailing-slash asymmetry** → Fixed in c74dcf0: extracted a `sameUrl(a, b)` helper that compares via `new URL(a).href === new URL(b).href`, with a raw-string fallback on parse errors. Added a test that pins `"https://x"` vs `"https://x/"` collapses to the same URL.
- **`ENV_KEYS` missing fallback variants** → Fixed in c74dcf0: added `ENVIO_RPC_FALLBACK_URL_11142220` and `ENVIO_RPC_FALLBACK_URL_10143` to the snapshot/restore list.

552 tests pass (was 551 + 1 new trailing-slash test), typecheck + lint clean.

File: indexer-envio/src/rpc/client.ts [hunk 1]
```diff
@@ -255,30 +257,51 @@ export function getRpcClient(
...
-  if (!config) return null;
+  if (!config) {
+    fallbackRpcClients.set(chainId, null);
+    return null;
+  }
 
-  const fallbackUrl = withHyperRpcToken(config.default);
-  if (isBareHyperRpcUrl(fallbackUrl)) return null;
+  const fallbackOverride = process.env[`ENVIO_RPC_FALLBACK_URL_${chainId}`];
+  const fallbackRawUrl = fallbackOverride ?? config.default;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/client.ts]: Fixed in 73baf71 — `ENVIO_RPC_FALLBACK_URL_<chainId>` now treats empty string as unset (matches the existing primary-URL truthy check). Added a test pinning the behaviour. Good catch.

File: indexer-envio/.env.example [hunk 2]
```diff
@@ -4,15 +4,25 @@
...
+# ENVIO_RPC_URL_42220="https://forno.celo.org"          # Celo Mainnet primary (default)
+# ENVIO_RPC_URL_11142220="<celo-sepolia-rpc>"            # Celo Sepolia primary (default: forno)
+# ENVIO_RPC_URL_143="https://rpc2.monad.xyz"             # Monad Mainnet primary (default — deeper archive than QuickNode)
+# ENVIO_RPC_URL_10143="<monad-testnet-full-node-rpc>"    # Monad Testnet primary (default: HyperRPC — needs ENVIO_API_TOKEN)
+
+# --- Per-chain fallback RPC overrides (optional) ---
+# Used by readContractWithBlockFallback for archive-depth + rate-limit failover.
+# Set when neither the primary's archive depth nor its rate limit covers the
+# full sync window — e.g. on Monad keep primary = rpc2.monad.xyz (deep
+# archive) and set fallback = a tokenized QuickNode URL (high rate limit).
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/.env.example]: Partial fix in 73baf71. The structural concern is valid — `readContractWithBlockFallback` uses one URL for both archive-depth AND rate-limit failover, so a shallow secondary leaks rate-limit failures into archive-depth misses on deep blocks during catch-up.

Won't fix the structural side in this PR (deferring it): the proper fix is splitting archive-depth and rate-limit fallback into separate URLs (or making rate-limit fallback block-depth aware), which is a larger redesign. For this PR I:

- Tightened the `.env.example` + `AGENTS.md` recommendation to make the trade-off explicit: the swap pattern only works when the deep-archive primary rarely rate-limits at the indexer's load. In our specific Monad case, `rpc2.monad.xyz` is the official Monad RPC and we don't see rate-limit pressure at our event volume — but the docs no longer present this as universally safe.
- Filed the structural fix as a follow-up.

In the meantime, the failure mode you describe (rpc2 rate-limits → QuickNode archive-depth-misses → null) results in the same caller behaviour as any other RPC failure: the entity update is skipped for that event, and the next event triggers another fetch attempt. Not a permanent state loss, just additional re-tries until the primary rate-limit clears.

File: indexer-envio/src/rpc/client.ts [hunk 3]
```diff
@@ -255,30 +257,51 @@ export function getRpcClient(
...
+  const fallbackUrl = withHyperRpcToken(fallbackRawUrl);
+  if (isBareHyperRpcUrl(fallbackUrl)) {
+    fallbackRpcClients.set(chainId, null);
+    return null;
+  }
 
   const perChainOverride = process.env[config.envVar];
   const legacyGlobal = process.env.ENVIO_RPC_URL;
-  if (!perChainOverride && !legacyGlobal) return null;
+  const primaryRawUrl = perChainOverride ?? legacyGlobal ?? config.default;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply indexer-envio/src/rpc/client.ts]: Fixed in f0da341 — switched primary URL resolution to use `||` (truthiness) instead of `??` (nullish coalescing) so it matches `getRpcClient`'s `if (perChainOverride)`/`if (legacyGlobal)` semantics. Empty-string env vars now fall through consistently in both functions, eliminating the self-referencing-fallback case. Added a regression test.
